File: app/src/main/java/com/zack/enderplan/view/activity/AboutActivity.java
package com.zack.enderplan.view.activity;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.design.widget.AppBarLayout;
import android.support.v4.view.ViewPager;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;
import android.view.ViewTreeObserver;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.pixelcan.inkpageindicator.InkPageIndicator;
import com.zack.enderplan.App;
import com.zack.enderplan.R;
import com.zack.enderplan.injector.component.DaggerAboutComponent;
import com.zack.enderplan.injector.module.AboutPresenterModule;
import com.zack.enderplan.presenter.AboutPresenter;
import com.zack.enderplan.view.adapter.AboutPagerAdapter;
import com.zack.enderplan.view.contract.AboutViewContract;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;

public class AboutActivity extends BaseActivity implements AboutViewContract {

    @BindView(R.id.layout_app_bar)
    AppBarLayout mAppBarLayout;
    @BindView(R.id.toolbar)
    Toolbar mToolbar;
    @BindView(R.id.layout_header)
    LinearLayout mHeaderLayout;
    @BindView(R.id.text_version)
    TextView mVersionText;
    @BindView(R.id.pager_about)
    ViewPager mAboutPager;
    @BindView(R.id.indicator_about)
    InkPageIndicator mAboutIndicator;

    @Inject
    AboutPresenter mAboutPresenter;

    public static void start(Context context) {
        context.startActivity(new Intent(context, AboutActivity.class));
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mAboutPresenter.attach();
    }

    @Override
    protected void onInjectPresenter() {
        DaggerAboutComponent.builder()
                .aboutPresenterModule(new AboutPresenterModule(this, getSupportFragmentManager()))
                .appComponent(App.getAppComponent())
                .build()
                .inject(this);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        mAboutPresenter.detach();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onBackPressed() {
        mAboutPresenter.notifyBackPressed();
    }

    @Override
    public void showInitialView(String versionName, AboutPagerAdapter aboutPagerAdapter) {
        setContentView(R.layout.activity_about);
        ButterKnife.bind(this);

        setSupportActionBar(mToolbar);
        setupActionBar();

        mAppBarLayout.getViewTreeObserver().addOnPreDrawListener(new ViewTreeObserver.OnPreDrawListener() {
            @Override
            public boolean onPreDraw() {
                mAppBarLayout.getViewTreeObserver().removeOnPreDrawListener(this);
                mAboutPresenter.notifyPreDrawingAppBar(mAppBarLayout.getTotalScrollRange());
                return false;
            }
        });

        mAppBarLayout.addOnOffsetChangedListener(new AppBarLayout.OnOffsetChangedListener() {
            @Override
            public void onOffsetChanged(AppBarLayout appBarLayout, int verticalOffset) {
                mAboutPresenter.notifyAppBarScrolled(verticalOffset);
            }
        });

        mVersionText.setText(versionName);
        mAboutPager.setAdapter(aboutPagerAdapter);
        mAboutIndicator.setViewPager(mAboutPager);
    }

    @Override
    public void onAppBarScrolled(float headerLayoutAlpha) {
        mHeaderLayout.setAlpha(headerLayoutAlpha);
    }

    @Override
    public void onAppBarScrolledToCriticalPoint(String toolbarTitle) {
        mToolbar.setTitle(toolbarTitle);
    }

    @Override
    public void backToTop() {
        mAppBarLayout.setExpanded(true);
    }

    @Override
    public void pressBack() {
        super.onBackPressed();
    }
}

File: app/src/main/java/net/zackzhang/app/end/view/activity/AboutActivity.kt
package net.zackzhang.app.end.view.activity

import android.animation.ObjectAnimator
import android.content.Context
import android.content.Intent
import android.graphics.Path
import android.hardware.SensorManager
import android.os.Bundle
import android.support.design.widget.AppBarLayout
import android.view.MenuItem
import android.view.ViewTreeObserver
import kotlinx.android.synthetic.main.activity_about.*
import kotlinx.android.synthetic.main.content_about.*
import net.zackzhang.app.end.App
import net.zackzhang.app.end.R
import net.zackzhang.app.end.injector.component.DaggerAboutComponent
import net.zackzhang.app.end.injector.module.AboutPresenterModule
import net.zackzhang.app.end.presenter.AboutPresenter
import net.zackzhang.app.end.view.adapter.AboutPagerAdapter
import net.zackzhang.app.end.view.contract.AboutViewContract
import javax.inject.Inject

class AboutActivity : BaseActivity(), AboutViewContract {

    companion object {

        fun start(context: Context) {
            context.startActivity(Intent(context, AboutActivity::class.java))
        }
    }

    @Inject
    lateinit var mAboutPresenter: AboutPresenter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mAboutPresenter.attach()
    }

    override fun onInjectPresenter() {
        DaggerAboutComponent.builder()
                .aboutPresenterModule(AboutPresenterModule(this, supportFragmentManager, getSystemService(Context.SENSOR_SERVICE) as SensorManager))
                .appComponent(App.appComponent)
                .build()
                .inject(this)
    }

    override fun onResume() {
        super.onResume()
        mAboutPresenter.notifyRegisteringSensorListener()
    }

    override fun onPause() {
        super.onPause()
        mAboutPresenter.notifyUnregisteringSensorListener()
    }

    override fun onStop() {
        super.onStop()
        //将位置恢复放在这，就看不到移动的过程了
        mAboutPresenter.notifyResetingViewTranslation()
    }

    override fun onDestroy() {
        super.onDestroy()
        mAboutPresenter.detach()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> finish()
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onBackPressed() {
        mAboutPresenter.notifyBackPressed()
    }

    override fun showInitialView(versionName: String, aboutPagerAdapter: AboutPagerAdapter) {
        setContentView(R.layout.activity_about)

        setSupportActionBar(toolbar)
        setupActionBar()

        //注释掉这一句使AppBar可折叠
        (layout_collapsing_toolbar.layoutParams as AppBarLayout.LayoutParams).scrollFlags = 0

        layout_app_bar.viewTreeObserver.addOnPreDrawListener(object : ViewTreeObserver.OnPreDrawListener {
            override fun onPreDraw(): Boolean {
                layout_app_bar.viewTreeObserver.removeOnPreDrawListener(this)
                mAboutPresenter.notifyPreDrawingAppBar(layout_app_bar.totalScrollRange)
                return false
            }
        })

        layout_app_bar.addOnOffsetChangedListener(AppBarLayout.OnOffsetChangedListener { _, verticalOffset -> mAboutPresenter.notifyAppBarScrolled(verticalOffset) })

        text_version.text = versionName
        pager_about.adapter = aboutPagerAdapter
        indicator_about.setViewPager(pager_about)
    }

    override fun onAppBarScrolled(headerLayoutAlpha: Float) {
        layout_header.alpha = headerLayoutAlpha
    }

    override fun onAppBarScrolledToCriticalPoint(toolbarTitle: String) {
        toolbar.title = toolbarTitle
    }

    override fun translateViewWhenIncline(shouldTranslateX: Boolean, translationX: Float, shouldTranslateY: Boolean, translationY: Float) {
        if (!shouldTranslateX && !shouldTranslateY) return
        val path = Path()
        //start point
        path.moveTo(layout_header.translationX, layout_header.translationY)
        //end point
        if (shouldTranslateX && shouldTranslateY) {
            //x和y均合适
            path.lineTo(translationX, translationY)
        } else if (shouldTranslateX) {
            //x合适，y不合适
            path.lineTo(translationX, layout_header.translationY)
        } else {
            //x不合适，y合适
            path.lineTo(layout_header.translationX, translationY)
        }
        ObjectAnimator.ofFloat(layout_header, "translationX", "translationY", path).setDuration(80).start()
    }

    override fun resetViewTranslation() {
        layout_header.translationX = 0f
        layout_header.translationY = 0f
    }

    override fun backToTop() {
        layout_app_bar.setExpanded(true)
    }

    override fun pressBack() {
        super.onBackPressed()
    }
}
